### PR TITLE
ICU-21551 Fixing unit prefix normalization order to match LDML spec

### PR DESCRIPTION
**Pull request by younies**
Fixes https://github.com/icu-units/icu/issues/154 

##### Checklist

- [x] Required: Issue filed: https://unicode-org.atlassian.net/browse/ICU-21551
- [x] Required: The PR title must be prefixed with a JIRA Issue number. 
- [x] Required: The PR description must include the link to the Jira Issue, for example by completing the URL in the first checklist item
- [x] Required: Each commit message must be prefixed with a JIRA Issue number. 
- [x] Issue accepted (done by Technical Committee after discussion)
- [x] Tests included, if applicable
- [ ] API docs and/or User Guide docs changed or added, if applicable

### CODE REVIEW

File: icu4c/source/test/intltest/measfmttest.cpp [hunk 1]
```diff
@@ -4472,12 +4472,7 @@ void MeasureFormatTest::TestIdentifiers() {
         {"exbibyte", "exbibyte"},
         {"zebibyte", "zebibyte"},
         {"yobibyte", "yobibyte"},
-
-        // Testing sort order of prefixes.
-        //
-        // TODO(icu-units#70): revisit when fixing normalization. For now we're
-        // just checking some consistency between C&J.
-        {"megafoot-mebifoot-kibifoot-kilofoot", "kibifoot-mebifoot-kilofoot-megafoot"},
+        {"megafoot-mebifoot-kibifoot-kilofoot", "kilofoot-kibifoot-megafoot-mebifoot"},
```
Comment (hugovdm) [inline icu4c/source/test/intltest/measfmttest.cpp:4475]: This is not spec compliant - "sort the larger prefixes first" - flip the comparator? (https://www.unicode.org/reports/tr35/tr35-info.html#Unit_Identifier_Normalization)

**Review by hugovdm — CHANGES_REQUESTED** [review]
Other than reversing the sort order, this looks good to me. Add the same change to Java?

File: icu4c/source/test/intltest/measfmttest.cpp [hunk 2]
```diff
@@ -4472,12 +4472,10 @@ void MeasureFormatTest::TestIdentifiers() {
...
         {"yobibyte", "yobibyte"},
-
         // Testing sort order of prefixes.
-        //
-        // TODO(icu-units#70): revisit when fixing normalization. For now we're
-        // just checking some consistency between C&J.
-        {"megafoot-mebifoot-kibifoot-kilofoot", "kibifoot-mebifoot-kilofoot-megafoot"},
+        {"megafoot-mebifoot-kibifoot-kilofoot", "mebifoot-megafoot-kibifoot-kilofoot"},
+        // Testing sort of units.
+        {"inch-foot-yard", "foot-inch-yard"}, // TODO: the result should be "yard-foot-inch"
```
Comment (hugovdm) [inline icu4c/source/test/intltest/measfmttest.cpp:4478]: Are you sure? I don't think there's anything in the specification regarding the order of yard, foot, and inch.

(Spec: https://www.unicode.org/reports/tr35/tr35-info.html#Unit_Identifier_Normalization)

If I were you, I would treat it as out of scope for this pull request, and file a CLDR ticket for clarification of the specification.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/units/SingleUnitImpl.java [hunk 3]
```diff
@@ -122,21 +122,49 @@ int compareTo(SingleUnitImpl other) {
...
+        // When comparing binary prefixes vs SI prefixes, instead of comparing the actual values, we can
+        // multiply the binary prefix power by 3 and compare the powers. if they are equal, we can can
+        // compare the bases.
+        // NOTE: this methodology will fail if the binary prefix more than or equal 98.
+        int unitBase = this.unitPrefix.getBase();
+        int otherUnitBase = other.unitPrefix.getBase();
+        // Values for comparison purposes only.
+        int unitPowerCompAbs =
+                unitBase == 1024 /* Binary Prefix */ ? Math.abs(this.unitPrefix.getPower()) * 3
+                        : Math.abs(this.unitPrefix.getPower());
```
Comment (hugovdm) [inline icu4j/main/classes/core/src/com/ibm/icu/impl/units/SingleUnitImpl.java:135]: As with C++, I don't think taking absolute values makes sense?

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java [hunk 4]
```diff
@@ -3923,12 +3923,11 @@ public void TestIdentifiers() {
...
             new TestCase("yobibyte", "yobibyte"),
-
             // Testing sort order of prefixes.
-            //
-            // TODO(icu-units#70): revisit when fixing normalization. For now we're
-            // just checking some consistency between C&J.
-            new TestCase("megafoot-mebifoot-kibifoot-kilofoot", "kibifoot-mebifoot-kilofoot-megafoot"),
+            new TestCase("megafoot-mebifoot-kibifoot-kilofoot", "mebifoot-megafoot-kibifoot-kilofoot"),
+            // Testing sort of units.
+            new TestCase("inch-foot-yard", "foot-inch-yard"), // TODO: the result should be "yard-foot-inch"
```
Comment (hugovdm) [inline icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java:3929]: As with C++, the TODO doesn't look spec-compliant to me, I think the sort order is undefined.

Also: add negative exponents tests (e.g. milli, micro, nano).

**Review by hugovdm — CHANGES_REQUESTED** [review]

File: icu4c/source/i18n/measunit_impl.h [hunk 5]
```diff
@@ -105,22 +106,60 @@ struct U_I18N_API SingleUnitImpl : public UMemory {
...
+
+        // Values for comparison purposes only.
+        int32_t unitPowerCompAbs =
+            unitBase == 1024 /* Binary Prefix */ ? std::abs(umeas_getPrefixPower(unitPrefix)) * 3
+                                                 : std::abs(umeas_getPrefixPower(unitPrefix));
+        int32_t otherUnitPowerCompAbs =
+            otherUnitBase == 1024 /* Binary Prefix */ ? std::abs(
+                                                            umeas_getPrefixPower(other.unitPrefix)) *
+                                                            3
+                                                      : std::abs(umeas_getPrefixPower(other.unitPrefix));
```
[the remark this answers is not in the document]
Comment (younies) [reply icu4c/source/i18n/measunit_impl.h:133]: I have added a new test for illustration
```
        {"megafoot-mebifoot-kibifoot-kilofoot", "mebifoot-megafoot-kibifoot-kilofoot"},
        {"per-megafoot-mebifoot-kibifoot-kilofoot", "per-mebifoot-megafoot-kibifoot-kilofoot"},
        {"megafoot-mebifoot-kibifoot-kilofoot-per-megafoot-mebifoot-kibifoot-kilofoot",
         "mebifoot-megafoot-kibifoot-kilofoot-per-mebifoot-megafoot-kibifoot-kilofoot"},
```

I do not think we need to flip the order  after `per`

File: icu4c/source/test/intltest/measfmttest.cpp (hunk 2 above)
Comment (younies) [reply icu4c/source/test/intltest/measfmttest.cpp:4478]: okay, good idea.

File: icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java (hunk 4 above)
Comment (younies) [reply icu4j/main/tests/core/src/com/ibm/icu/dev/test/format/MeasureUnitTest.java]: done

File: icu4c/source/test/intltest/measfmttest.cpp [hunk 6]
```diff
@@ -4472,12 +4472,10 @@ void MeasureFormatTest::TestIdentifiers() {
         {"exbibyte", "exbibyte"},
         {"zebibyte", "zebibyte"},
         {"yobibyte", "yobibyte"},
-
         // Testing sort order of prefixes.
-        //
-        // TODO(icu-units#70): revisit when fixing normalization. For now we're
-        // just checking some consistency between C&J.
-        {"megafoot-mebifoot-kibifoot-kilofoot", "kibifoot-mebifoot-kilofoot-megafoot"},
+        {"megafoot-mebifoot-kibifoot-kilofoot", "mebifoot-megafoot-kibifoot-kilofoot"},
```
[the remark this answers is not in the document]
Comment (younies) [reply icu4c/source/test/intltest/measfmttest.cpp:4476]: done

**Comment by hugovdm** [comment]
When I try to select "Show changes since your last review", I get an error:
"We went looking everywhere, but couldn’t find those commits."

I would suggest when you have sent something for review at a particular commit, you don't force-push that commit away, such that the reviewer can more easily compare what's been changed since.

**Comment by hugovdm** [comment]
Looks like I can fetch an old commit:

$ git fetch units
$ git fetch units c5378e823e5d99476f392da3a0aab1db3e6f372a
$ git diff c5378e823e5d99476f392da3a0aab1db3e6f372a units/master

**Comment by hugovdm** [comment]
Looks like this has been rebased since the last review. This also makes reviewing harder. I'll restore:

$ git checkout c5378e823e5d99476f392da3a0aab1db3e6f372a
$ git checkout -b fix-ordering-hugovdm
$ git rebase upstream/main
$ git diff fix-ordering-hugovdm units/fix-ordering

That's looking cleaner. Reviewing!

File: icu4c/source/test/intltest/measfmttest.cpp (hunk 1 above)
Comment (hugovdm) [reply icu4c/source/test/intltest/measfmttest.cpp]: This looks done. Thanks!

**Comment by hugovdm** [comment]
Getting to the actual commit I think I reviewed at:

$ git fetch units 86624422d5afa334602dc54cd98adee457454084
$ git tree 86624422d5afa334602dc54cd98adee457454084 --all  # Looks good
$ git checkout 86624422d5afa334602dc54cd98adee457454084
$ git checkout -B fix-ordering-hugovdm
$ git rebase upstream/main
$ git diff fix-ordering-hugovdm units/fix-ordering

**Comment by hugovdm** [comment]
Even after this, I still haven't found the "std::abs(...)" that I commented on previously. It's thus been hard to focus on this, but I think I've managed to pull myself together: the abs() is gone. Back to the review!

File: icu4c/source/i18n/measunit_impl.h (hunk 5 above)
[the remark this answers is not in the document]
Comment (hugovdm) [reply icu4c/source/i18n/measunit_impl.h]: Thanks for the new unit tests! They look good to me.

And I see the abs is gone. The C++ code looks good to me.

File: icu4j/main/classes/core/src/com/ibm/icu/impl/units/SingleUnitImpl.java (hunk 3 above)
Comment (hugovdm) [reply icu4j/main/classes/core/src/com/ibm/icu/impl/units/SingleUnitImpl.java]: I had forgotten I had even reviewed some Java code. :blush: But I've already spent more time on this than I can afford before perf. Can I pass this on to @sffc to review the Java code / ensuring it matches C++?

**Comment by hugovdm** [comment]
Assigning to @sffc to perhaps review the Java code before I'm able to get back to it: I can't do that today.

File: icu4c/source/i18n/measunit_impl.h [hunk 7]
```diff
@@ -105,22 +106,47 @@ struct U_I18N_API SingleUnitImpl : public UMemory {
...
+
+        // Values for comparison purposes only.
+        int32_t unitPower = unitBase == 1024 /* Binary Prefix */ ? umeas_getPrefixPower(unitPrefix) * 3
+                                                                 : umeas_getPrefixPower(unitPrefix);
+        int32_t otherUnitPower =
+            otherUnitBase == 1024 /* Binary Prefix */ ? umeas_getPrefixPower(other.unitPrefix) * 3
+                                                      : umeas_getPrefixPower(other.unitPrefix);
+
+        // NOTE: if the unitPower is less than the other,
+        // we return 1 not -1. Thus because we want th sorting order
```
Comment (younies) [inline icu4c/source/i18n/measunit_impl.h:133]: th/the

**Review by sffc — DISMISSED** [review]
J and C look equivalent.  Test cases look good.

**Comment by younies** [comment]
@sffc , I squashed the branch, therefore, I need another approval.

**Review by sffc — DISMISSED** [review]

**Comment by younies** [comment]
@sffc , just fixing merging conflict and squashed

**Review by sffc — APPROVED** [review]

**Pull request merged**